pkg/volume/csi/csi_attacher_test.go

```diff
@@ -1630,7 +1607,30 @@ func newTestWatchPlugin(t *testing.T, fakeClient *fakeclient.Clientset, setupInf
 		t.Fatalf("cannot assert plugin to be type csiPlugin")
 	}
 
-	return csiPlug, fakeWatcher, tmpDir, fakeClient
+	attachmentWatchCreated, fakeWatcher := notifyVolumeAttachmentWatchCreated(fakeClient)
+
+	return csiPlug, attachmentWatchCreated, fakeWatcher, tmpDir, fakeClient
+}
+
+// notifyVolumeAttachmentWatchCreated returns a channel that is closed once a VolumeAttachment watch is created.
+// This is useful for ensuring a watch has been created prior to injecting events.
+func notifyVolumeAttachmentWatchCreated(client *fakeclient.Clientset) (chan core.Action, *watch.RaceFreeFakeWatcher) {
+	attachmentWatchCreated := make(chan core.Action)
+	fakeWatcher := watch.NewRaceFreeFake()
```

Comment:
The default watch reactor creates a new fake watch for each call to `.Watch`. Sharing one among (potentially) multiple callers will mean those callers all end up reading from the same channel, so only one reader will get each event.

Reply:
Yes, I encountered issues with this. I'm not sure what the best approach is given I need a way of signaling that the watch is created so I can begin injecting events (which is the purpose of this channel we return to the caller).

Reply:
Closing the channel and then deferring to the default reactor would have the same effect, wouldn't it? It seems like some of the tests expect to interact directly with the fake watcher -- is that how they inject events? Are these events that can't be injected through normal `client.Foo.Bar.Create()` calls?

Reply:
Deferring to the default reactor was the previous behavior, but then you would encounter race conditions where you'd see the following:

1. First reactor runs, closes channel, signifying watch created
2. Test injects event into watcher
3. Second reactor runs, returns watcher to client
4. Client receives watcher and never sees event (because it already happened)

This comment points out similar behavior:
https://github.com/kubernetes/kubernetes/blob/9406e145c25e4c3041ec82529176957f164fdb08/pkg/volume/csi/csi_attacher_test.go#L367-L372

I wasn't able to successfully inject events through the API and see them in the watcher. I had to interact directly with the watcher to trigger events. I'm not sure where I'm going wrong that's preventing this from working.

Reply:
It seems like I need a better mechanism for signaling the watch is ready, because right now I'm prepending watch reactors to the client after the shared informer factory has started, and this is hitting data races between `PrependWatchReactor()` and `InvokesWatch()`. It seems I need to add these reactors prior to starting the factory.

Reply:
I notice when I remove the prepended reactors (including the `core.DefaultWatchReactor(fakeWatcher, nil)`), changes made at via the API are reflected in the watcher.

Reply:
Data race addressed here: https://github.com/kubernetes/kubernetes/pull/97584